### PR TITLE
feat: use supertest for server testing 🔥

### PR DESCRIPTION
**Pull request by nguyenvukhang**
### Summary of changes

- use supertest for server testing 🔥🔥🔥
  - directory path matches api call route
  - test filenames are methods (there's a test at `meta.test.ts` to ensure that this rule is followed)
- rely on server's `app.ts` to handle all responses 🔥🔥🔥
  - api methods no longer need to all do the `res.json` then-catch dance

### Testing

- Logout, login and search for a module, place it down and place another down
- `yarn build` and `yarn test` from root as usual

### CODE REVIEW

File: apps/server/src/api/user.ts [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+        .then(flatten.user)
+    }
+
+  /**
+   * lists all Users
+   *
+   * @param {Api} api
+   */
+  static list = (api: Api) => async (req: CustomReqQuery<ListRequest>) => {
+    const { id, authZeroId, email } = req.query
```
Comment (weiseng18) [inline apps/server/src/api/user.ts:56]: might want to use `req.body` for email

File: apps/server/test/degrees/[DELETE].test.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+import { setup, teardown } from '@modtree/test-env'
+import { Api } from '@modtree/repo-api'
+
+const dbName = oneUp(__filename)
+const db = getSource(dbName)
+let app: Express
+let api: Api
+let findOneById: jest.SpyInstance
+let remove: jest.SpyInstance
+
```
Comment (weiseng18) [inline apps/server/test/degrees/[DELETE].test.ts:15]: what's the difference between this test and `/degrees/:degreeId/[DELETE].test.ts`

File: apps/server/test/graphs/:graphId/flow/[POST].test.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+afterAll(() => teardown(db))
+
+// dd9f2725-7a08-45f8-bbbb-10bb7e002bb3
+// 017fc011-486c-4ec6-a038-9a92ab85a8f3
+
+const testRequest = async () =>
+  request(app).post('/graphs/017fc011-486c-4ec6-a038-9a92ab85a8f3/flow').send({
+    flowNodes: [],
+    flowEdges: [],
+  })
```
Comment (weiseng18) [inline apps/server/test/graphs/:graphId/flow/[POST].test.ts:32]: where do we initialize the entities in the server tests

File: apps/server/src/routes.ts [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+} from './api'
+import { body, param, query } from 'express-validator'
+import { validModuleCode } from '@modtree/utils'
+
+export const routes: Route[] = [
+  /**
+   * User API
+   */
+  {
+    method: 'get',
```
Comment (weiseng18) [inline apps/server/src/routes.ts:17]: should be `POST`

File: apps/server/src/api/user.ts (hunk 1 above)
Comment (nguyenvukhang) [reply apps/server/src/api/user.ts:56]: get requests don't take body parameters

File: apps/server/src/routes.ts (hunk 4 above)
Comment (nguyenvukhang) [reply apps/server/src/routes.ts:17]: it's actually `GET`
https://auth0.com/docs/api/management/v2#!/Users/get_users

File: apps/server/src/api/user.ts (hunk 1 above)
Comment (weiseng18) [reply apps/server/src/api/user.ts:56]: I'm not sure if email works as a query param. Also we might not want to put it as a query param for privacy

File: apps/server/src/routes.ts (hunk 4 above)
Comment (weiseng18) [reply apps/server/src/routes.ts:17]: I guess if we aren't using the body then it can be a GET

File: apps/server/src/routes.ts [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    route: '/users',
+    fn: UserApi.list,
+    validators: [
+      body('id').isUUID().optional(),
+      body('email').normalizeEmail().isEmail().optional(),
+      body('authZeroId')
+        .custom((s: string) => s.startsWith('auth0|'))
+        .optional(),
+    ],
```
Comment (weiseng18) [inline apps/server/src/routes.ts:26]: Change to query validation if still using GET

**Review by weiseng18 — APPROVED** [review]

**Pull request merged**